crates/extricrate/src/lib.rs

```diff
@@ -44,13 +64,174 @@ pub mod dependencies {
+        ModuleDoesNotExists(String),
+    }
+
+    pub fn get_crate_entrypoint(crate_root: &Path) -> Option<PathBuf> {
+        // TODO: support custom entry point
+
+        let cargo_toml = crate_root.join("Cargo.toml");
+        if !cargo_toml.exists() {
+            return None;
+        }
+
+        let main_rs = crate_root.join(Path::new("src/main.rs"));
+        if main_rs.exists() {
+            return Some(main_rs);
+        }
+
+        let lib_rs = crate_root.join(Path::new("src/lib.rs"));
+        if lib_rs.exists() {
+            return Some(lib_rs);
+        };
+        None
+    }
+
     /// List all the `use` statements in the crate, by file/module.
-    pub fn list_use_statements(crate_root: &std::path::Path) -> UseStatementMap {
-        todo!()
+    pub fn list_use_statements(
+        crate_root: &Path,
+    ) -> Result<UseStatementMap, ListUseStatementError> {
+        let mut files_visited = HashSet::new();
+        let mut files_to_visit = VecDeque::new();
+        let entry_point =
+            get_crate_entrypoint(crate_root).ok_or(ListUseStatementError::PathIsNotACrate)?;
+        let src_folder = entry_point
+            .parent()
+            .expect("Failed to get entry point parent folder");
+        files_to_visit.push_back(entry_point.clone());
+        while let Some(file_to_visit) = files_to_visit.pop_front() {
+            if files_visited.contains(&file_to_visit) {
+                continue;
+            }
+
+            if !file_to_visit.exists() {
+                return Err(ListUseStatementError::FileNotFound);
+            }
+
+            let content = read_to_string(&file_to_visit)
+                .map_err(|_| ListUseStatementError::FileNotReadable)?;
+
+            let parsed_file =
+                parse_file(&content).map_err(|_| ListUseStatementError::FileNotParsable)?;
+
+            let mut visitor = UseVisitor::new();
+            visitor.visit_file(&parsed_file);
+            for dependency in visitor.dependencies {
+                match dependency {
+                    UseStatementType::Simple(name) => {
+                        if is_local_import(&name) {
+                            let file_to_visit = get_path_from_module_name(src_folder, &name)
+                                .ok_or(ListUseStatementError::ModuleDoesNotExists(name))?;
+                            files_to_visit.push_back(file_to_visit);
+                        }
+                    }
+                    UseStatementType::Alias(name, _) => todo!(),
+                    UseStatementType::WildCard(name) => todo!(),
+                }
+            }
+            files_visited.insert(file_to_visit);
+        }
+
+        todo!();
+    }
+
+    fn get_path_from_module_name(src_folder: &Path, name: &str) -> Option<PathBuf> {
+        let relative_path = name.strip_prefix("crate::").unwrap_or(name);
+        let relative_path = relative_path
+            .strip_prefix("my_crate::")
```

Comment:
I don't think there's anything special about "my_crate", it's just an example crate name, not a keyword

Reply:
totally, I'll add a todo to actually read the `Cargo.toml` crate name 👍